brave/brave/build.gradle

```diff
@@ -1,3 +1,10 @@
 dependencies {
-    api project(':brave5')
+    api project(':brave6')
+    // Brave 5.x depended on zipkin-reporter-brave, but Brave 6.x has no
+    // dependencies. We add back dependencies that allow ZipkinSpanHandler to
+    // compile. This allows people to migrate without managing these
+    // dependencies on their own. Those who want a dependency-free integration
+    // can use armeria-brave6.
+    api libs.zipkin.reporter.brave // For AsyncZipkinSpanHandler (native Brave)
+    api libs.zipkin // For ZipkinSpanHandler (uses Zipkin core types)
```

Comment:
Shouldn't we add this to brave6 module instead?

Reply:
`AsyncZipkinSpanHandler` is the only good way to report spans, `ZipkinSpanHandler` was a legacy bridge, and pins zipkin classes (rather than brave's native MutablSpan which has its own codecs)

Reply:
starting in brave 5, it didn't need a dependency on the library zipkin server needed. brave 6 removed the deprecated functions that still accessed those types. brave 6 can work with zipkin-reporter 2 (which uses zipkin core types) or zipkin-reporter 3 (which doesn't). If we pin brave 6 here to zipkin core types it adds a dependency that it several hundred K and won't be used. I highly suggest the LINE internal code stop using zipkin core types and use MutableSpan instead, or declare their own dependency on core types.

Reply:
I was told this was about metrics, so it is far more efficient for the internal code to skip entirely switching to zipkin core types. here's an example of how to do something like thathttps://github.com/openzipkin/brave/blob/master/brave/src/test/java/brave/features/handler/SpanMetricsCustomizer.java

Reply:
also, I will help port it if someone puts it public, really we don't want to pin deps for an edge case, especially as it took a lot of effort to remove this dep ;)

Reply:
In general I'm very curious about what armeria's expectations are about indirect dependencies.. that a project should forever declare dependencies on anything they depended on at some point, even after switching major version numbers?

I think at some point people should start to declare dependencies on the types they rely on in main code, that are indirectly supplied by another artifact. Is this strange? How does netty handle it? Has moving from 3 to 4 included every indirect dependencies 3 ever used?

Reply:
`armeria-brave` was was simply retained as an alias of `armeria-brave5` so I thought adding extra dependency declaration to the `build.gradle` doesn't make sense.
Have you thought about this comment?
https://github.com/line/armeria/pull/5453#pullrequestreview-1879249321

Reply:
> In general I'm very curious about what armeria's expectations are about indirect dependencies.. that a project should forever declare dependencies on anything they depended on at some point, even after switching major version numbers?

We don't usually declare dependencies unless there's a version conflict issue. ;) Again, I believe we just wanted to keep the `armeria-brave` module as an alias of `armeria-brave5` (now it's for `armeria-brave6`), so I thought it didn't make sense to add them. 😆

Reply:
I think armeria-brave5 is dead, as there's a near zero chance of any release on that version. Because the problem is about an indirect dep and LINE is already on armeria-brave (which is based on 6), I think removing both and keeping only brave6 is the best course, although it is a shame it can't just be armeria-brave

Reply:
put another way, while I added tests to ensure armeria-brave can compile random types also in zipkin, I feel it is wrong. I'm only doing this in deference to the major sponsor of this project. If keeping armeria-brave with the extra deps helps fund armeria. practical enough. OTOH it isn't needed.

What is important is to not mess up what's currently called brave6 by adding random deps to it which can re-introduce conflicts. If we do anyway, for example to get just one armeria-brave, it is fine, just let's put a comment loudly on when this will be removed and that its purpose is to help internal code move over to adding an explicit dep (even if via zipkin-reporter-bom) on types they use.

Reply:
> I think removing both and keeping only brave6 is the best course, although it is a shame it can't just be armeria-brave

Okay, let me talk to the team tomorrow once more if we can completely remove `armeria-brave5` or not. If it's not possible, you still want to retain `armeria-brave` module (with brave6). Is that correct?

Reply:
so after shower, here's an idea.

we *add an armeria sender* somewhere, and in that case someone can depend on that and get the correct recent reporter deps. If we add it to armeria-brave, basically that's fine and at least isn't random to add reporter (v3) anymore (because [sender](https://github.com/openzipkin/brave-example/blob/master/armeria/src/main/java/brave/example/WebClientSender.java) and [its tests](https://github.com/openzipkin/zipkin-reporter-java/tree/master/urlconnection/src/test/java/zipkin2/reporter/urlconnection) will need zipkin-reporter 3 types).

What we are left with is that large zipkin dep `io.zipkin.zipkin2:zipkin` which this was all about. zipkin-reporter-brave and all the senders exclude it because it is  huge and unused even when reporting to zipkin, if you use the most current way `AsyncZipkinSpanHandler` as our [armeria-example](https://github.com/openzipkin/brave-example/blob/master/armeria/src/main/java/brave/example/HttpTracingFactory.java) does.
```xml
    <dependency>
      <groupId>${project.groupId}</groupId>
      <artifactId>zipkin-reporter</artifactId>
      <version>${project.version}</version>
      <!-- Senders don't use zipkin types. Excluding allows brave users to
           avoid them by default. -->
      <exclusions>
        <exclusion>
          <groupId>io.zipkin.zipkin2</groupId>
          <artifactId>zipkin</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
```

So, basically it boils things down to one thing.. do we need to keep adding a dep on this? If so, we need to manage zipkin's version, and keep bumping like dependabot even though it isn't used, and also any app that uses brave will be a few hundred K larger than it needs to be.

That's all the info and pros and cons, so I'll leave you to decison making. Thanks for listening @minwoox!

Reply:
Let me answer your question.

> In general I'm very curious about what armeria's expectations are about indirect dependencies.. that a project should forever declare dependencies on anything they depended on at some point, even after switching major version numbers?

It depends. 😉 If the indirect dependencies are tightly coupled in terms of using our API, we want to retain them even after a major version update so that: 1) users can seamlessly upgrade the Armeria version without modifying their dependency configuration, and 2) we can manage the versions of indirect dependencies. If they're not coupled, we simply let users choose whatever they prefer.

> do we need to keep adding a dep on this? If so, we need to manage zipkin's version, and keep bumping like dependabot even though it isn't used,

So it's the same context here. If most of `armeria-brave` users use `AsyncZipkinSpanHandler`, we definitely need to add `zipkin-reporter`. Let me investigate how armeria-brave users use the API so that we can make a decision. 😉

Reply:
What's probably confusing is that in Brave 4 these were coupled, Brave 5 decoupled them but kept back compat, and Brave 6 removed coupling.

Adding zipkin-reporter-brave is fine, as it is small. The side effect of adding this is that we have to pin a specific version of zipkin-reporter (3.x now), and will have a problem again on 4.x. zipkin-reporter is a different library than brave, even if it has a bridge. For example, it is also used in otel.

regardless,  it won't bring in the polarizing and unneeded dep io.zipkin.zipkin2 dep.

So there are two decisions
 1. for convenience, pin reporter (as done now in armeria-brave) and know this needs to be revisited in the event of a reporter 4?
 2. if you do that, also add the pretty large io.zipkin2:zipkin dep

over to you!

Reply:
After discussing with the team, we've decided that removing the armeria-brave module is the best approach to address the indirect dependency problem and eliminate any ambiguity by maintaining three different modules.

Previously, we retained the armeria-brave module to avoid causing breaking changes for users. However, we've now agreed that removing the armeria-brave module and providing armeria-brave5 is acceptable, as users will not need to modify their code. They only need to change the module name in the build file.

By removing armeria-brave, which has dependencies on zipkin, we can resolve the indirect dependency issue with armeria-brave6, which only includes the brave dependency.

Furthermore, we plan to deprecate the `armeria-brave5` module by adding `@Deprecated` to the `package-info.java` file within the module. This will facilitate users in migrating to `armeria-brave6`.

Please, let me know if we have any missing points and what you think!
Thanks, Adrian!

Reply:
sounds like a good plan, I didn't think about package-info.java for deprecation.. that's a neat trick I'll add to my bag 👍